File: cli/src/libindy/ledger.rs
use indy::IndyError;
use indy::future::Future;
use indy::ledger;

pub struct Ledger {}

impl Ledger {
    pub fn sign_and_submit_request(pool_handle: i32, wallet_handle: i32, submitter_did: &str, request_json: &str) -> Result<String, IndyError> {
        ledger::sign_and_submit_request(pool_handle, wallet_handle, submitter_did, request_json).wait()
    }

    pub fn submit_request(pool_handle: i32, request_json: &str) -> Result<String, IndyError> {
        ledger::submit_request(pool_handle, request_json).wait()
    }

    pub fn submit_action(pool_handle: i32, request_json: &str, nodes: Option<&str>, timeout: Option<i32>) -> Result<String, IndyError> {
        ledger::submit_action(pool_handle, request_json, nodes, timeout).wait()
    }

    pub fn sign_request(wallet_handle: i32, submitter_did: &str, request_json: &str) -> Result<String, IndyError> {
        ledger::sign_request(wallet_handle, submitter_did, request_json).wait()
    }

    pub fn multi_sign_request(wallet_handle: i32, submitter_did: &str, request_json: &str) -> Result<String, IndyError> {
        ledger::multi_sign_request(wallet_handle, submitter_did, request_json).wait()
    }

    pub fn build_nym_request(submitter_did: &str, target_did: &str, verkey: Option<&str>,
                             data: Option<&str>, role: Option<&str>) -> Result<String, IndyError> {
        ledger::build_nym_request(submitter_did, target_did, verkey, data, role).wait()
    }

    pub fn build_get_nym_request(submitter_did: Option<&str>, target_did: &str) -> Result<String, IndyError> {
        ledger::build_get_nym_request(submitter_did, target_did).wait()
    }

    pub fn build_attrib_request(submitter_did: &str, target_did: &str, hash: Option<&str>, raw: Option<&str>, enc: Option<&str>) -> Result<String, IndyError> {
        ledger::build_attrib_request(submitter_did, target_did, hash, raw, enc).wait()
    }

    pub fn build_get_attrib_request(submitter_did: Option<&str>, target_did: &str, raw: Option<&str>, hash: Option<&str>, enc: Option<&str>) -> Result<String, IndyError> {
        ledger::build_get_attrib_request(submitter_did, target_did, raw, hash, enc).wait()
    }

    pub fn build_schema_request(submitter_did: &str, data: &str) -> Result<String, IndyError> {
        ledger::build_schema_request(submitter_did, data).wait()
    }

    pub fn build_get_schema_request(submitter_did: Option<&str>, id: &str) -> Result<String, IndyError> {
        ledger::build_get_schema_request(submitter_did, id).wait()
    }

    pub fn build_cred_def_request(submitter_did: &str, data: &str) -> Result<String, IndyError> {
        ledger::build_cred_def_request(submitter_did, data).wait()
    }

    pub fn build_get_validator_info_request(submitter_did: &str) -> Result<String, IndyError> {
        ledger::build_get_validator_info_request(submitter_did).wait()
    }

    pub fn build_get_cred_def_request(submitter_did: Option<&str>, id: &str) -> Result<String, IndyError> {
        ledger::build_get_cred_def_request(submitter_did, id).wait()
    }

    pub fn build_node_request(submitter_did: &str, target_did: &str, data: &str) -> Result<String, IndyError> {
        ledger::build_node_request(submitter_did, target_did, data).wait()
    }

    pub fn indy_build_pool_config_request(submitter_did: &str, writes: bool, force: bool) -> Result<String, IndyError> {
        ledger::build_pool_config_request(submitter_did, writes, force).wait()
    }

    pub fn indy_build_pool_restart_request(submitter_did: &str, action: &str, datetime: Option<&str>) -> Result<String, IndyError> {
        ledger::build_pool_restart_request(submitter_did, action, datetime).wait()
    }

    pub fn indy_build_pool_upgrade_request(submitter_did: &str, name: &str, version: &str, action: &str, sha256: &str, timeout: Option<u32>, schedule: Option<&str>,
                                           justification: Option<&str>, reinstall: bool, force: bool, package: Option<&str>) -> Result<String, IndyError> {
        ledger::build_pool_upgrade_request(submitter_did, name, version, action, sha256,
                                           timeout, schedule, justification,
                                           reinstall, force, package).wait()
    }

    pub fn build_auth_rule_request(submitter_did: &str,
                                   txn_type: &str,
                                   action: &str,
                                   field: &str,
                                   old_value: Option<&str>,
                                   new_value: Option<&str>,
                                   constraint: &str, ) -> Result<String, IndyError> {
        ledger::build_auth_rule_request(submitter_did, txn_type, action, field,
                                        old_value, new_value, constraint).wait()
    }

    pub fn build_auth_rules_request(submitter_did: &str,
                                    rules: &str, ) -> Result<String, IndyError> {
        ledger::build_auth_rules_request(submitter_did, rules).wait()
    }

    pub fn build_get_auth_rule_request(submitter_did: Option<&str>,
                                       auth_type: Option<&str>,
                                       auth_action: Option<&str>,
                                       field: Option<&str>,
                                       old_value: Option<&str>,
                                       new_value: Option<&str>, ) -> Result<String, IndyError> {
        ledger::build_get_auth_rule_request(submitter_did, auth_type, auth_action, field,
                                            old_value, new_value).wait()
    }

    pub fn build_txn_author_agreement_request(submitter_did: &str, text: &str, version: &str) -> Result<String, IndyError> {
        ledger::build_txn_author_agreement_request(submitter_did, text, version).wait()
    }

    pub fn build_acceptance_mechanisms_request(submitter_did: &str, aml: &str, version: &str, aml_context: Option<&str>) -> Result<String, IndyError> {
        ledger::build_acceptance_mechanisms_request(submitter_did, aml, version, aml_context).wait()
    }

    pub fn build_get_txn_author_agreement_request(submitter_did: Option<&str>,
                                                  data: Option<&str>, ) -> Result<String, IndyError> {
        ledger::build_get_txn_author_agreement_request(submitter_did, data).wait()
    }

    pub fn append_txn_author_agreement_acceptance_to_request(request_json: &str,
                                                             text: Option<&str>,
                                                             version: Option<&str>,
                                                             hash: Option<&str>,
                                                             acc_mech_type: &str,
                                                             time_of_acceptance: u64) -> Result<String, IndyError> {
        ledger::append_txn_author_agreement_acceptance_to_request(request_json,
                                                                  text,
                                                                  version,
                                                                  hash,
                                                                  acc_mech_type,
                                                                  time_of_acceptance).wait()
    }

    pub fn append_request_endorser(request_json: &str,
                                   endorser_did: &str) -> Result<String, IndyError> {
        ledger::append_request_endorser(request_json, endorser_did).wait()
    }
}

File: wrappers/nodejs/test/ledger.js
var test = require('ava')
var indy = require('../')
var cuid = require('cuid')
var initTestPool = require('./helpers/initTestPool')
var tempy = require('tempy')

function sleep (ms) {
  return new Promise(function (resolve) {
    setTimeout(resolve, ms)
  })
}

async function waitUntilApplied (ph, req, cond) {
  for (let i = 0; i < 3; i++) {
    let res = await indy.submitRequest(ph, req)

    if (cond(res)) {
      return res
    }

    await sleep(5 * 1000)
  }
}

test('ledger', async function (t) {
  var pool = await initTestPool()
  var walletConfig = { 'id': 'wallet-' + cuid() }
  var walletCredentials = { 'key': 'key' }
  await indy.createWallet(walletConfig, walletCredentials)
  var wh = await indy.openWallet(walletConfig, walletCredentials)
  var [trusteeDid] = await indy.createAndStoreMyDid(wh, { seed: '000000000000000000000000Trustee1' })
  var [myDid, myVerkey] = await indy.createAndStoreMyDid(wh, { })
  var schemaName = 'schema-' + cuid()
  var [schemaId, schema] = await indy.issuerCreateSchema(myDid, schemaName, '1.0', ['name', 'age'])

  // Nym
  var req = await indy.buildNymRequest(trusteeDid, myDid, myVerkey, null, 'TRUSTEE')
  var res = await indy.signAndSubmitRequest(pool.handle, wh, trusteeDid, req)
  t.is(res.result.txn.data.verkey, myVerkey)

  var resMetadata = await indy.getResponseMetadata(res)
  t.true(resMetadata.hasOwnProperty('seqNo'))
  t.true(resMetadata.hasOwnProperty('txnTime'))
  t.false(resMetadata.hasOwnProperty('lastTxnTime'))
  t.false(resMetadata.hasOwnProperty('lastSeqNo'))

  req = await indy.buildGetNymRequest(trusteeDid, myDid)
  t.is(req.identifier, trusteeDid)
  t.is(req.operation.dest, myDid)

  // Schema
  req = await indy.buildSchemaRequest(myDid, schema)
  req = await indy.signRequest(wh, myDid, req)
  res = await indy.submitRequest(pool.handle, req)

  req = await indy.buildGetSchemaRequest(myDid, schemaId)
  res = await waitUntilApplied(pool.handle, req, res => res['result']['seqNo'] != null)
  var data = await indy.parseGetSchemaResponse(res)
  t.is(data[0], schemaId)
  t.is(data[1].name, schema.name)
  req = await indy.buildGetTxnRequest(myDid, null, data[1].seqNo)
  res = await waitUntilApplied(pool.handle, req, res => res['result']['data']['txnMetadata']['seqNo'] != null)
  t.is(res.result.data.txn.data.data.name, schema.name)
  schema = data[1]

  // Node
  req = await indy.buildNodeRequest(myDid, myDid, {
    node_ip: '10.0.0.100',
    node_port: 9710,
    client_ip: '10.0.0.100',
    client_port: 9709,
    alias: 'Node5',
    services: ['VALIDATOR'],
    blskey: 'CnEDk9HrMnmiHXEV1WFgbVCRteYnPqsJwrTdcZaNhFVW'
  })
  res = await indy.submitRequest(pool.handle, req)
  t.is(res.op, 'REQNACK')

  // Attrib
  req = await indy.buildAttribRequest(myDid, myDid, null, { endpoint: { ha: '127.0.0.1:5555' } }, null)
  res = await indy.signAndSubmitRequest(pool.handle, wh, myDid, req)

  req = await indy.buildGetAttribRequest(myDid, myDid, 'endpoint', null, null)
  res = await waitUntilApplied(pool.handle, req, data => data['result']['data'] != null)
  t.deepEqual(JSON.parse(res.result.data), { endpoint: { ha: '127.0.0.1:5555' } })

  // Pool
  req = await indy.buildPoolConfigRequest(myDid, false, false)
  t.false(req.operation.writes)

  req = await indy.buildPoolRestartRequest(myDid, 'start', '0')
  t.is(req.operation.action, 'start')

  req = await indy.buildPoolUpgradeRequest(myDid, 'some upgrade action', '2.0.0', 'cancel', 'abc12345', -1, null, null, false, false, null)
  t.is(req.operation.name, 'some upgrade action')

  // DDO
  req = await indy.buildGetDdoRequest(myDid, trusteeDid)
  t.is(req.operation.dest, trusteeDid)

  // Cred Def
  var [credDefId, credDef] = await indy.issuerCreateAndStoreCredentialDef(wh, myDid, schema, 'TAG', 'CL', { support_revocation: true })
  req = await indy.buildCredDefRequest(myDid, credDef)
  res = await indy.signAndSubmitRequest(pool.handle, wh, myDid, req)

  req = await indy.buildGetCredDefRequest(myDid, credDefId)
  res = await waitUntilApplied(pool.handle, req, res => res['result']['seqNo'] != null)
  res = await indy.parseGetCredDefResponse(res)
  t.is(res[0], credDefId)
  t.is(res[1].id, credDef.id)

  // Revoc Reg Def
  var writerH = await indy.openBlobStorageWriter('default', {
    'base_dir': tempy.directory(),
    'uri_pattern': ''
  })
  var [revRegDefId, revRegDef, revRegEntry] = await indy.issuerCreateAndStoreRevocReg(wh, myDid, null, 'tag1', credDefId, { max_cred_num: 5 }, writerH)

  req = await indy.buildRevocRegDefRequest(myDid, revRegDef)
  res = await indy.signAndSubmitRequest(pool.handle, wh, myDid, req)
  t.is(res.result.txn.data.id, revRegDefId)

  req = await indy.buildGetRevocRegDefRequest(myDid, revRegDefId)
  res = await waitUntilApplied(pool.handle, req, res => res['result']['seqNo'] != null)
  res = await indy.parseGetRevocRegDefResponse(res)
  t.is(res[0], revRegDefId)
  t.is(res[1].id, revRegDef.id)

  // Revoc Reg Entry
  req = await indy.buildRevocRegEntryRequest(myDid, revRegDefId, 'CL_ACCUM', revRegEntry)
  res = await indy.signAndSubmitRequest(pool.handle, wh, myDid, req)

  var nowSeconds = Math.floor(Date.now() / 1000)
  req = await indy.buildGetRevocRegRequest(myDid, revRegDefId, nowSeconds + 100)
  res = await waitUntilApplied(pool.handle, req, res => res['result']['seqNo'] != null)
  res = await indy.parseGetRevocRegResponse(res)
  t.is(res[0], revRegDefId)
  t.is(typeof res[1], 'object')
  t.is(typeof res[2], 'number')

  // RevocRegDelta
  req = await indy.buildGetRevocRegDeltaRequest(myDid, revRegDefId, nowSeconds, nowSeconds + 100)
  res = await waitUntilApplied(pool.handle, req, res => res['result']['seqNo'] != null)
  res = await indy.parseGetRevocRegDeltaResponse(res)
  t.is(res[0], revRegDefId)
  t.is(typeof res[1], 'object')
  t.is(typeof res[2], 'number')

  // Submit Action
  req = await indy.buildGetValidatorInfoRequest(myDid)
  req = await indy.signRequest(wh, myDid, req)
  res = await indy.submitAction(pool.handle, req, null, null)

  // Auth Rule
  req = await indy.buildGetAuthRuleRequest(trusteeDid, 'NYM', 'ADD', 'role', null, '101')
  res = await indy.submitRequest(pool.handle, req)
  var defaultConstraint = res['result']['data'][0]['constraint']

  var constraint = {
    'sig_count': 1,
    'metadata': {},
    'role': '0',
    'constraint_id': 'ROLE',
    'need_to_be_owner': false
  }
  req = await indy.buildAuthRuleRequest(trusteeDid, 'NYM', 'ADD', 'role', null, '101', constraint)
  res = await indy.signAndSubmitRequest(pool.handle, wh, trusteeDid, req)
  t.is(res.op, 'REPLY')

  await sleep(1000)

  req = await indy.buildGetAuthRuleRequest(trusteeDid, 'NYM', 'ADD', 'role', null, '101')
  res = await indy.submitRequest(pool.handle, req)
  t.deepEqual(res['result']['data'][0]['constraint'], constraint)

  var expectedAuthRule = {
    'auth_type': 'NYM',
    'auth_action': 'ADD',
    'field': 'role',
    'new_value': '101',
    'constraint': constraint
  }

  var authRulesData = [expectedAuthRule]
  req = await indy.buildAuthRulesRequest(trusteeDid, authRulesData)
  res = await indy.submitRequest(pool.handle, req)
  t.deepEqual(req['operation'], { 'type': '122', 'rules': authRulesData })

  // author agreement
  req = await indy.buildTxnAuthorAgreementRequest(trusteeDid, 'indy agreement', '1.0.0')
  t.deepEqual(req['operation'], { 'type': '4', 'text': 'indy agreement', 'version': '1.0.0' })

  req = await indy.buildGetTxnAuthorAgreementRequest(null, { 'version': '1.0.0' })
  t.deepEqual(req['operation'], { 'type': '6', 'version': '1.0.0' })

  // acceptance mechanism
  var aml = { 'acceptance mechanism label 1': 'some acceptance mechanism description 1' }
  req = await indy.buildAcceptanceMechanismsRequest(trusteeDid, aml, '1.0.0', null)
  t.deepEqual(req['operation'], { 'type': '5', 'aml': aml, 'version': '1.0.0' })

  req = await indy.buildGetAcceptanceMechanismsRequest(null, 123456789, null)
  t.deepEqual(req['operation'], { 'type': '7', 'timestamp': 123456789 })

  // author agreement acceptance data
  req = await indy.appendTxnAuthorAgreementAcceptanceToRequest(req, 'indy agreement', '1.0.0', null, 'acceptance mechanism label 1', 123456789)
  var expectedMeta = {
    'mechanism': 'acceptance mechanism label 1',
    'taaDigest': '7213b9aabf8677edf6b17d20a9fbfaddb059ea4cb122d163bdf658ea67196120',
    'time': 123456789
  }
  t.deepEqual(req['taaAcceptance'], expectedMeta)

  // set back
  req = await indy.buildAuthRuleRequest(trusteeDid, 'NYM', 'ADD', 'role', null, '101', defaultConstraint)
  res = await indy.signAndSubmitRequest(pool.handle, wh, trusteeDid, req)
  t.is(res.op, 'REPLY')

  await indy.closeWallet(wh)
  await indy.deleteWallet(walletConfig, walletCredentials)
  pool.cleanup()
})
